src/v/kafka/server/handlers/describe_transactions.cc

```diff
@@ -0,0 +1,141 @@
+// Copyright 2020 Redpanda Data, Inc.
+//
+// Use of this software is governed by the Business Source License
+// included in the file licenses/BSL.md
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0
+
+#include "kafka/server/handlers/describe_transactions.h"
+
+#include "cluster/errc.h"
+#include "cluster/tx_gateway_frontend.h"
+#include "kafka/protocol/errors.h"
+#include "kafka/protocol/schemata/describe_transactions_request.h"
+#include "kafka/protocol/types.h"
+#include "kafka/server/handlers/details/security.h"
+#include "kafka/server/request_context.h"
+#include "kafka/server/response.h"
+#include "model/fundamental.h"
+#include "model/namespace.h"
+#include "resource_mgmt/io_priority.h"
+
+#include <seastar/core/coroutine.hh>
+#include <seastar/core/when_all.hh>
+
+#include <unordered_map>
+
+namespace kafka {
+
+namespace {
+
+ss::future<> fill_info_about_tx(
+  cluster::tx_gateway_frontend& tx_frontend,
+  describe_transactions_response& response,
+  kafka::transactional_id tx_id) {
+    describe_transaction_state tx_info_resp;
+    tx_info_resp.transactional_id = tx_id;
+
+    auto tx_info = co_await tx_frontend.get_tx(tx_id);
+    if (tx_info.has_value()) {
+        auto tx = tx_info.value();
+        tx_info_resp.producer_id = kafka::producer_id(tx.pid.id);
+        tx_info_resp.producer_epoch = tx.pid.get_epoch();
+        tx_info_resp.transaction_state = ss::sstring(tx.get_kafka_status());
+        tx_info_resp.transaction_timeout_ms = tx.get_timeout().count();
+        // RP doesn't store transaction_start_time so we use last_update
+        // instead
+        tx_info_resp.transaction_start_time_ms
+          = tx.last_update_ts.time_since_epoch().count();
```

Comment:
I think I am not sure about time stuff. Because in out admin_api we return diff. But here we need to return real timestamp. But we do not use normal time from std

Reply:
in order to make it a kafka timestamp it must be milliseconds since linux epoch please cast the time_since_epoch to ms and then call `count()`

Reply:
In test I found 
```
       new ProducerState()
                .setProducerId(12345L)
                .setProducerEpoch(15)
                .setLastSequence(75)
                .setLastTimestamp(System.currentTimeMillis())
                .setCurrentTxnStartOffset(-1L),
```

So they uss milliseconds